File: go/periph/chips/imu/mpu9250.go
package imu

import (
	"fmt"
	"time"

	"github.com/tuhde/Periph/go/periph/connection"
)

// MPU-9250 register map (subset; shared with MPU-6050 plus magnetometer
// access path).
const (
	reg9250SmplrtDiv   = 0x19
	reg9250Config      = 0x1A
	reg9250GyroConfig  = 0x1B
	reg9250AccelConfig = 0x1C
	reg9250AccelConfig2 = 0x1D
	reg9250FifoEn      = 0x23
	reg9250IntPinCfg   = 0x37
	reg9250IntStatus   = 0x3A
	reg9250AccelXoutH  = 0x3B
	reg9250TempOutH    = 0x41
	reg9250GyroXoutH   = 0x43
	reg9250UserCtrl    = 0x6A
	reg9250PwrMgmt1    = 0x6B
	reg9250FifoCountH  = 0x72
	reg9250FifoR_W     = 0x74
	reg9250WhoAmI      = 0x75

	whoAmI9250Value = 0x71
)

// AK8963 magnetometer registers (address 0x0C).
const (
	ak8963WIA   = 0x00
	ak8963ST1   = 0x02
	ak8963HXL   = 0x03
	ak8963ST2   = 0x09
	ak8963CNTL1 = 0x0A
	ak8963CNTL2 = 0x0B
	ak8963ASAX  = 0x10
	ak8963ASAY  = 0x11
	ak8963ASAZ  = 0x12

	ak8963IDExpected = 0x48

	ak8963ModePowerDown = 0x00
	ak8963ModeFuseROM   = 0x0F
	ak8963Cont100Hz16b = 0x16
)

const (
	mpu9250ResetDelay        = 100 * time.Millisecond
	mpu9250GyroStartupDelay  = 35 * time.Millisecond
	mpu9250MagPowerDownDelay = 10 * time.Millisecond
)

// MPU9250Minimal is the MPU-9250 9-axis MotionTracking device driver
// (3-axis accelerometer + 3-axis gyroscope) — minimal interface. No
// magnetometer in Minimal; see MPU9250Full for AK8963 access.
//
// Performs device reset, WHO_AM_I verification, and configures sensible
// defaults at construction. Default configuration baked in:
//   - Gyroscope full-scale: ±250 dps (GYRO_FS_SEL=0)
//   - Accelerometer full-scale: ±2 g (ACCEL_FS_SEL=0)
//   - Gyro DLPF: 41 Hz bandwidth (DLPF_CFG=3)
//   - Accel DLPF: 42 Hz bandwidth (A_DLPFCFG=3)
//   - Sample rate: 200 Hz (SMPLRT_DIV=4)
//   - Clock: auto PLL (CLKSEL=1)
type MPU9250Minimal struct {
	connection connection.Connection
	accelFs   uint8
	gyroFs    uint8
}

// NewMPU9250Minimal creates a new MPU9250Minimal, resets the device,
// verifies WHO_AM_I, and configures defaults at construction.
//
// connection must be a configured I²C connection bound to the device's
// 7-bit address (0x68 default, 0x69 alternate).
func NewMPU9250Minimal(t connection.Connection) (*MPU9250Minimal, error) {
	d := &MPU9250Minimal{connection: t}
	if err := d.writeReg(reg9250PwrMgmt1, 0x80); err != nil {
		return nil, err
	}
	time.Sleep(mpu9250ResetDelay)
	if err := d.writeReg(reg9250PwrMgmt1, 0x01); err != nil {
		return nil, err
	}
	who, err := d.readReg(reg9250WhoAmI)
	if err != nil {
		return nil, err
	}
	if who != whoAmI9250Value {
		return nil, fmt.Errorf("MPU9250 WHO_AM_I: expected 0x%02X, got 0x%02X", whoAmI9250Value, who)
	}
	if err := d.writeReg(reg9250GyroConfig, 0x00); err != nil {
		return nil, err
	}
	if err := d.writeReg(reg9250AccelConfig, 0x00); err != nil {
		return nil, err
	}
	if err := d.writeReg(reg9250AccelConfig2, 0x00); err != nil {
		return nil, err
	}
	if err := d.writeReg(reg9250Config, 0x03); err != nil {
		return nil, err
	}
	if err := d.writeReg(reg9250SmplrtDiv, 0x04); err != nil {
		return nil, err
	}
	time.Sleep(mpu9250GyroStartupDelay)
	return d, nil
}

func (d *MPU9250Minimal) writeReg(reg, value byte) error {
	return d.connection.Write([]byte{reg, value})
}

func (d *MPU9250Minimal) readReg(reg byte) (byte, error) {
	b, err := d.connection.WriteRead([]byte{reg}, 1)
	if err != nil {
		return 0, err
	}
	return b[0], nil
}

func (d *MPU9250Minimal) readReg16Signed(reg byte) (int16, error) {
	b, err := d.connection.WriteRead([]byte{reg}, 2)
	if err != nil {
		return 0, err
	}
	return int16(uint16(b[0])<<8 | uint16(b[1])), nil
}

// Accel reads the 3-axis linear acceleration.
//
// Returns (x, y, z) in m/s².
func (d *MPU9250Minimal) Accel() (float32, float32, float32, error) {
	b, err := d.connection.WriteRead([]byte{reg9250AccelXoutH}, 6)
	if err != nil {
		return 0, 0, 0, err
	}
	ax := int16(uint16(b[0])<<8 | uint16(b[1]))
	ay := int16(uint16(b[2])<<8 | uint16(b[3]))
	az := int16(uint16(b[4])<<8 | uint16(b[5]))
	sens := accelSensitivity[d.accelFs]
	return float32(ax) / sens * 9.80665,
		float32(ay) / sens * 9.80665,
		float32(az) / sens * 9.80665,
		nil
}

// Gyro reads the 3-axis angular rate.
//
// Returns (x, y, z) in rad/s.
func (d *MPU9250Minimal) Gyro() (float32, float32, float32, error) {
	b, err := d.connection.WriteRead([]byte{reg9250GyroXoutH}, 6)
	if err != nil {
		return 0, 0, 0, err
	}
	gx := int16(uint16(b[0])<<8 | uint16(b[1]))
	gy := int16(uint16(b[2])<<8 | uint16(b[3]))
	gz := int16(uint16(b[4])<<8 | uint16(b[5]))
	sens := gyroSensitivity[d.gyroFs]
	const piOver180 = float32(3.141592653589793) / 180.0
	return float32(gx) / sens * piOver180,
		float32(gy) / sens * piOver180,
		float32(gz) / sens * piOver180,
		nil
}

// MPU9250Full is the MPU-9250 9-axis MotionTracking device driver — full
// interface. Extends MPU9250Minimal with configuration, magnetometer
// access via the on-board AK8963, FIFO, and power control.
//
// The magnetometer is accessed via the chip's I²C bypass mode (BYPASS_EN
// in INT_PIN_CFG); the host can then address the AK8963 directly at
// 0x0C. To avoid forcing the caller to construct two connections with
// different addresses, NewMPU9250Full takes a connection factory that
// produces a new I²C connection bound to a caller-supplied 7-bit address
// on the same bus.
type MPU9250Full struct {
	*MPU9250Minimal
	magConnection connection.Connection
	magEnabled   bool
	magScaleX    float32
	magScaleY    float32
	magScaleZ    float32
	magBits      uint8
}

// ConnectionFactory returns a Connection bound to the given 7-bit address on
// the same underlying bus. Used by MPU9250Full to construct the secondary
// connection for the AK8963 magnetometer.
type ConnectionFactory func(addr uint8) (connection.Connection, error)

// NewMPU9250Full creates a new MPU9250Full with the same initialization as
// NewMPU9250Minimal. The magFactory is used to construct the AK8963
// connection on demand when EnableMag is called; it must return a
// connection bound to the given 7-bit address (0x0C for the AK8963).
func NewMPU9250Full(t connection.Connection, magFactory ConnectionFactory) (*MPU9250Full, error) {
	m, err := NewMPU9250Minimal(t)
	if err != nil {
		return nil, err
	}
	if magFactory == nil {
		return nil, fmt.Errorf("MPU9250Full: magFactory is required for AK8963 access")
	}
	mag, err := magFactory(0x0C)
	if err != nil {
		return nil, fmt.Errorf("MPU9250Full: open AK8963 connection: %w", err)
	}
	return &MPU9250Full{
		MPU9250Minimal: m,
		magConnection:   mag,
	}, nil
}

// Close releases the AK8963 secondary connection. The primary MPU9250
// connection is owned by the caller and not closed here.
func (d *MPU9250Full) Close() error {
	if d.magConnection != nil {
		return d.magConnection.Close()
	}
	return nil
}

// EnableMag initialises the on-board AK8963 magnetometer and switches it
// into continuous measurement mode.
//
// bits: 14 (BIT=0) or 16 (BIT=1).
// mode: 1=single, 2=8 Hz continuous, 6=100 Hz continuous.
func (d *MPU9250Full) EnableMag(bits, mode uint8) error {
	// 1) Enable I2C bypass so the host can address the AK8963 directly.
	if err := d.writeReg(reg9250IntPinCfg, 0x22); err != nil {
		return err
	}
	// 2) Power down the AK8963, wait 10 ms.
	if err := d.magConnection.Write([]byte{ak8963CNTL1, ak8963ModePowerDown}); err != nil {
		return err
	}
	time.Sleep(mpu9250MagPowerDownDelay)
	// 3) Enter fuse ROM access mode, wait 10 ms.
	if err := d.magConnection.Write([]byte{ak8963CNTL1, ak8963ModeFuseROM}); err != nil {
		return err
	}
	time.Sleep(mpu9250MagPowerDownDelay)
	// 4) Read ASAX/Y/Z factory calibration.
	asax, err := d.magConnection.WriteRead([]byte{ak8963ASAX}, 3)
	if err != nil {
		return err
	}
	d.magScaleX = (float32(asax[0]) - 128.0) / 256.0 + 1.0
	d.magScaleY = (float32(asax[1]) - 128.0) / 256.0 + 1.0
	d.magScaleZ = (float32(asax[2]) - 128.0) / 256.0 + 1.0
	// 5) Power down, wait 10 ms.
	if err := d.magConnection.Write([]byte{ak8963CNTL1, ak8963ModePowerDown}); err != nil {
		return err
	}
	time.Sleep(mpu9250MagPowerDownDelay)
	// 6) Configure continuous mode at the requested resolution.
	var bitFlag uint8
	if bits == 16 {
		bitFlag = 0x10
	}
	if err := d.magConnection.Write([]byte{ak8963CNTL1, bitFlag | (mode & 0x0F)}); err != nil {
		return err
	}
	d.magEnabled = true
	d.magBits = bits
	return nil
}

// Mag reads the 3-axis magnetic field strength.
//
// Returns (x, y, z) in µT. Mag must have been enabled first via EnableMag.
func (d *MPU9250Full) Mag() (float32, float32, float32, error) {
	if !d.magEnabled {
		return 0, 0, 0, fmt.Errorf("MPU9250 Mag: magnetometer not enabled (call EnableMag first)")
	}
	// Read 7 bytes from 0x03 (HXL) through 0x09 (ST2). ST2 must be read
	// to unlock the next measurement.
	b, err := d.magConnection.WriteRead([]byte{ak8963HXL}, 7)
	if err != nil {
		return 0, 0, 0, err
	}
	if b[6]&0x08 != 0 {
		return 0, 0, 0, fmt.Errorf("MPU9250 Mag: AK8963 overflow")
	}
	mx := int16(uint16(b[1])<<8 | uint16(b[0]))
	my := int16(uint16(b[3])<<8 | uint16(b[2]))
	mz := int16(uint16(b[5])<<8 | uint16(b[4]))
	var sens float32
	if d.magBits == 16 {
		sens = 0.15
	} else {
		sens = 0.6
	}
	return float32(mx) * sens * d.magScaleX,
		float32(my) * sens * d.magScaleY,
		float32(mz) * sens * d.magScaleZ,
		nil
}

// ConfigureGyro sets the gyroscope full-scale range.
func (d *MPU9250Full) ConfigureGyro(fullScale uint8) error {
	d.gyroFs = fullScale & 0x03
	return d.writeReg(reg9250GyroConfig, (fullScale&0x03)<<3)
}

// ConfigureAccel sets the accelerometer full-scale range.
func (d *MPU9250Full) ConfigureAccel(fullScale uint8) error {
	d.accelFs = fullScale & 0x03
	return d.writeReg(reg9250AccelConfig, (fullScale&0x03)<<3)
}

// ConfigureDLPF sets the gyroscope and accelerometer DLPF bandwidths.
func (d *MPU9250Full) ConfigureDLPF(gyroDLPF, accelDLPF uint8) error {
	if err := d.writeReg(reg9250Config, gyroDLPF&0x07); err != nil {
		return err
	}
	return d.writeReg(reg9250AccelConfig2, accelDLPF&0x07)
}

// ConfigureSampleRate sets the sample rate divider.
func (d *MPU9250Full) ConfigureSampleRate(divider uint8) error {
	return d.writeReg(reg9250SmplrtDiv, divider)
}

// Temperature reads the die temperature in °C.
func (d *MPU9250Full) Temperature() (float32, error) {
	raw, err := d.readReg16Signed(reg9250TempOutH)
	if err != nil {
		return 0, err
	}
	return float32(raw)/333.87 + 21.0, nil
}

// AccelRaw reads the raw 3-axis accelerometer values.
func (d *MPU9250Full) AccelRaw() (int16, int16, int16, error) {
	b, err := d.connection.WriteRead([]byte{reg9250AccelXoutH}, 6)
	if err != nil {
		return 0, 0, 0, err
	}
	return int16(uint16(b[0])<<8 | uint16(b[1])),
		int16(uint16(b[2])<<8 | uint16(b[3])),
		int16(uint16(b[4])<<8 | uint16(b[5])),
		nil
}

// GyroRaw reads the raw 3-axis gyroscope values.
func (d *MPU9250Full) GyroRaw() (int16, int16, int16, error) {
	b, err := d.connection.WriteRead([]byte{reg9250GyroXoutH}, 6)
	if err != nil {
		return 0, 0, 0, err
	}
	return int16(uint16(b[0])<<8 | uint16(b[1])),
		int16(uint16(b[2])<<8 | uint16(b[3])),
		int16(uint16(b[4])<<8 | uint16(b[5])),
		nil
}

// MagRaw reads the raw magnetometer values.
func (d *MPU9250Full) MagRaw() (int16, int16, int16, error) {
	if !d.magEnabled {
		return 0, 0, 0, fmt.Errorf("MPU9250 MagRaw: magnetometer not enabled (call EnableMag first)")
	}
	b, err := d.magConnection.WriteRead([]byte{ak8963HXL}, 7)
	if err != nil {
		return 0, 0, 0, err
	}
	return int16(uint16(b[1])<<8 | uint16(b[0])),
		int16(uint16(b[3])<<8 | uint16(b[2])),
		int16(uint16(b[5])<<8 | uint16(b[4])),
		nil
}

// DataReady returns true when RAW_DATA_RDY_INT is set in INT_STATUS.
func (d *MPU9250Full) DataReady() (bool, error) {
	s, err := d.readReg(reg9250IntStatus)
	if err != nil {
		return false, err
	}
	return s&0x01 != 0, nil
}

// SetSleep sets or clears the SLEEP bit in PWR_MGMT_1.
func (d *MPU9250Full) SetSleep(sleep bool) error {
	v, err := d.readReg(reg9250PwrMgmt1)
	if err != nil {
		return err
	}
	if sleep {
		v |= 0x40
	} else {
		v &^= 0x40
	}
	return d.writeReg(reg9250PwrMgmt1, v)
}

// FIFOcount returns the number of bytes in the FIFO.
func (d *MPU9250Full) FIFOcount() (uint16, error) {
	b, err := d.connection.WriteRead([]byte{reg9250FifoCountH}, 2)
	if err != nil {
		return 0, err
	}
	return (uint16(b[0])&0x1F)<<8 | uint16(b[1]), nil
}

// ReadFIFO reads up to len(buf) bytes from the FIFO.
func (d *MPU9250Full) ReadFIFO(buf []byte) (uint16, error) {
	count, err := d.FIFOcount()
	if err != nil {
		return 0, err
	}
	if count == 0 {
		return 0, nil
	}
	toRead := int(count)
	if toRead > len(buf) {
		toRead = len(buf)
	}
	read, err := d.connection.WriteRead([]byte{reg9250FifoR_W}, toRead)
	if err != nil {
		return 0, err
	}
	n := copy(buf, read)
	return uint16(n), nil
}

// EnableFIFO configures and enables FIFO sources.
func (d *MPU9250Full) EnableFIFO(gyro, accel, temp bool) error {
	var v uint8
	if accel {
		v |= 1 << 3
	}
	if temp {
		v |= 1 << 2
	}
	if gyro {
		v |= 1 << 4
	}
	if err := d.writeReg(reg9250FifoEn, v); err != nil {
		return err
	}
	uc, err := d.readReg(reg9250UserCtrl)
	if err != nil {
		return err
	}
	return d.writeReg(reg9250UserCtrl, uc|0x40)
}

// ResetFIFO resets the FIFO buffer.
func (d *MPU9250Full) ResetFIFO() error {
	uc, err := d.readReg(reg9250UserCtrl)
	if err != nil {
		return err
	}
	return d.writeReg(reg9250UserCtrl, uc|0x04)
}

File: python/periph/chips/imu/mpu9250.py
import struct
import time


class MPU9250Minimal:
    """MPU-9250 9-axis MotionTracking device (accelerometer + gyroscope) — minimal interface.

    Provides 3-axis acceleration and 3-axis angular rate readings with no
    configuration beyond the connection. Performs device reset, WHO_AM_I check,
    and enables all sensors at defaults during initialization. Magnetometer
    is not included in Minimal — it requires a separate initialization path.

    Default configuration (baked in at construction):
        - Gyroscope full-scale: ±250 dps (GYRO_FS_SEL=0)
        - Accelerometer full-scale: ±2 g (ACCEL_FS_SEL=0)
        - Gyroscope DLPF: 41 Hz bandwidth (CONFIG DLPF_CFG=3)
        - Accelerometer DLPF: 42 Hz bandwidth (ACCEL_CONFIG2 A_DLPFCFG=3)
        - Sample rate: 200 Hz (SMPLRT_DIV=4)
        - Clock: auto PLL (CLKSEL=1)
        - All six axes enabled
        - SPI only: I2C_IF_DIS set to prevent accidental I²C re-enable

    Args:
        connection: Configured I²C or SPI connection pointing at the device.
    """

    _REG_SMPLRT_DIV    = 0x19
    _REG_CONFIG        = 0x1A
    _REG_GYRO_CONFIG   = 0x1B
    _REG_ACCEL_CONFIG  = 0x1C
    _REG_ACCEL_CONFIG2 = 0x1D
    _REG_LP_ACCEL_ODR  = 0x1E
    _REG_WOM_THR       = 0x1F
    _REG_FIFO_EN       = 0x23
    _REG_INT_PIN_CFG   = 0x37
    _REG_INT_ENABLE    = 0x38
    _REG_INT_STATUS    = 0x3A
    _REG_ACCEL_XOUT_H  = 0x3B
    _REG_TEMP_OUT_H    = 0x41
    _REG_GYRO_XOUT_H   = 0x43
    _REG_USER_CTRL     = 0x6A
    _REG_PWR_MGMT_1    = 0x6B
    _REG_PWR_MGMT_2    = 0x6C
    _REG_FIFO_COUNTH   = 0x72
    _REG_FIFO_COUNTL   = 0x73
    _REG_FIFO_R_W      = 0x74
    _REG_WHO_AM_I      = 0x75

    _WHO_AM_I_VALUE = 0x71

    _ACCEL_SENSITIVITY = (16384.0, 8192.0, 4096.0, 2048.0)
    _GYRO_SENSITIVITY  = (131.0, 65.5, 32.8, 16.4)

    def __init__(self, connection):
        self._connection = connection
        self._accel_fs = 0
        self._gyro_fs = 0
        self._write_reg(self._REG_PWR_MGMT_1, 0x80)
        time.sleep(0.1)
        self._write_reg(self._REG_PWR_MGMT_1, 0x01)
        who = self._read_reg(self._REG_WHO_AM_I)
        if who != self._WHO_AM_I_VALUE:
            raise ValueError('MPU9250 WHO_AM_I: expected 0x{:02X}, got 0x{:02X}'.format(
                self._WHO_AM_I_VALUE, who))
        self._write_reg(self._REG_GYRO_CONFIG, 0x00)
        self._write_reg(self._REG_ACCEL_CONFIG, 0x00)
        self._write_reg(self._REG_ACCEL_CONFIG2, 0x00)
        self._write_reg(self._REG_CONFIG, 0x03)
        self._write_reg(self._REG_SMPLRT_DIV, 0x04)
        time.sleep(0.035)

    def _write_reg(self, reg, value):
        self._connection.write(bytes([reg, value]))

    def _read_reg(self, reg):
        return self._connection.write_read(bytes([reg]), 1)[0]

    def _read_reg16_signed(self, reg):
        raw = self._connection.write_read(bytes([reg]), 2)
        return struct.unpack('>h', raw)[0]

    def _read_burst(self, reg, n):
        return self._connection.write_read(bytes([reg]), n)

    def accel(self):
        """Read 3-axis linear acceleration.

        Returns:
            tuple: (x, y, z) acceleration in m/s².
        """
        raw = self._read_burst(self._REG_ACCEL_XOUT_H, 6)
        ax, ay, az = struct.unpack('>hhh', raw)
        sens = self._ACCEL_SENSITIVITY[self._accel_fs]
        return (ax / sens * 9.80665, ay / sens * 9.80665, az / sens * 9.80665)

    def gyro(self):
        """Read 3-axis angular rate.

        Returns:
            tuple: (x, y, z) angular rate in rad/s.
        """
        raw = self._read_burst(self._REG_GYRO_XOUT_H, 6)
        gx, gy, gz = struct.unpack('>hhh', raw)
        sens = self._GYRO_SENSITIVITY[self._gyro_fs]
        deg_to_rad = 3.141592653589793 / 180.0
        return (gx / sens * deg_to_rad,
                gy / sens * deg_to_rad,
                gz / sens * deg_to_rad)


class MPU9250Full(MPU9250Minimal):
    """MPU-9250 full interface — extends MPU9250Minimal with complete functionality.

    Adds gyroscope and accelerometer full-scale configuration, DLPF settings,
    sample rate control, temperature reading, magnetometer (AK8963) support,
    raw data access, data-ready polling, sleep/standby control, and FIFO management.

    Args:
        connection: Configured I²C or SPI connection pointing at the device.
    """

    _AK8963_ADDR = 0x0C

    _AK8963_REG_WIA      = 0x00
    _AK8963_REG_ST1      = 0x02
    _AK8963_REG_HXL      = 0x03
    _AK8963_REG_ST2      = 0x09
    _AK8963_REG_CNTL1    = 0x0A
    _AK8963_REG_CNTL2    = 0x0B
    _AK8963_REG_ASAX     = 0x10
    _AK8963_REG_ASAY     = 0x11
    _AK8963_REG_ASAZ     = 0x12

    _AK8963_WIA_VALUE = 0x48

    _MAG_SENSITIVITY_14BIT = 0.6
    _MAG_SENSITIVITY_16BIT = 0.15

    _GYRO_FS_SEL_BITS  = {0: 0, 1: 1, 2: 2, 3: 3}
    _ACCEL_FS_SEL_BITS = {0: 0, 1: 1, 2: 2, 3: 3}

    _CONFIG_DLPF = {0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6}
    _ACCEL_CONFIG2_DLPF = {0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6}

    def __init__(self, connection):
        super().__init__(connection)
        self._mag_enabled = False
        self._mag_bits = 16
        self._mag_scale_x = 1.0
        self._mag_scale_y = 1.0
        self._mag_scale_z = 1.0
        self._is_spi = False

    def _ak8963_write(self, reg, value):
        self._connection.write(bytes([self._AK8963_ADDR << 1, reg, value]))

    def _ak8963_read(self, reg):
        return self._connection.write_read(bytes([(self._AK8963_ADDR << 1) | 1, reg]), 1)[0]

    def _ak8963_read_burst(self, reg, n):
        return self._connection.write_read(bytes([(self._AK8963_ADDR << 1) | 1, reg]), n)

    def configure_gyro(self, full_scale=0):
        """Set gyroscope full-scale range.

        Args:
            full_scale: Range selector 0–3 (0=±250, 1=±500, 2=±1000, 3=±2000 dps).
        """
        self._gyro_fs = full_scale & 0x03
        self._write_reg(self._REG_GYRO_CONFIG, (full_scale & 0x03) << 3)

    def configure_accel(self, full_scale=0):
        """Set accelerometer full-scale range.

        Args:
            full_scale: Range selector 0–3 (0=±2g, 1=±4g, 2=±8g, 3=±16g).
        """
        self._accel_fs = full_scale & 0x03
        self._write_reg(self._REG_ACCEL_CONFIG, (full_scale & 0x03) << 3)

    def configure_dlpf(self, gyro_dlpf=3, accel_dlpf=3):
        """Set digital low-pass filter bandwidth.

        Args:
            gyro_dlpf: Gyro filter setting 0–6 (0=256 Hz, 1=188 Hz, 2=98 Hz,
                3=41 Hz, 4=20 Hz, 5=10 Hz, 6=5 Hz).
            accel_dlpf: Accel filter setting 0–6 (0=460 Hz, 1=184 Hz, 2=92 Hz,
                3=42 Hz, 5=20 Hz, 6=10 Hz). Note: 4 is not valid for accel.
        """
        cfg_val = gyro_dlpf & 0x07
        self._write_reg(self._REG_CONFIG, cfg_val)
        accel2_val = accel_dlpf & 0x07
        self._write_reg(self._REG_ACCEL_CONFIG2, accel2_val)

    def configure_sample_rate(self, divider=4):
        """Set sample rate divider.

        Args:
            divider: SMPLRT_DIV value 0–255; output rate = 1 kHz / (1 + divider)
                when DLPF is active.
        """
        self._write_reg(self._REG_SMPLRT_DIV, divider & 0xFF)

    def temperature(self):
        """Read die temperature.

        Returns:
            float: Temperature in °C.
        """
        raw = self._read_reg16_signed(self._REG_TEMP_OUT_H)
        return raw / 333.87 + 21.0

    def enable_mag(self, bits=16, mode=6):
        """Initialize AK8963 magnetometer via I²C bypass mode.

        Args:
            bits: Output resolution, 14 or 16.
            mode: Operation mode (1=single, 2=8 Hz continuous, 6=100 Hz continuous).
        """
        self._write_reg(self._REG_INT_PIN_CFG, 0x22)
        time.sleep(0.01)

        self._ak8963_write(self._AK8963_REG_CNTL1, 0x00)
        time.sleep(0.01)

        self._ak8963_write(self._AK8963_REG_CNTL1, 0x0F)
        time.sleep(0.01)

        asax = self._ak8963_read(self._AK8963_REG_ASAX)
        asay = self._ak8963_read(self._AK8963_REG_ASAY)
        asaz = self._ak8963_read(self._AK8963_REG_ASAZ)

        self._mag_scale_x = (asax - 128) / 256.0 + 1.0
        self._mag_scale_y = (asay - 128) / 256.0 + 1.0
        self._mag_scale_z = (asaz - 128) / 256.0 + 1.0

        self._ak8963_write(self._AK8963_REG_CNTL1, 0x00)
        time.sleep(0.01)

        cntl1_val = 0
        if bits == 16:
            cntl1_val |= 0x10
        cntl1_val |= (mode & 0x0F)
        self._ak8963_write(self._AK8963_REG_CNTL1, cntl1_val)
        time.sleep(0.01)

        self._mag_enabled = True
        self._mag_bits = bits

    def mag(self):
        """Read 3-axis magnetic field.

        Returns:
            tuple: (x, y, z) magnetic field in µT.

        Raises:
            RuntimeError: If magnetometer has not been enabled via enable_mag().
        """
        if not self._mag_enabled:
            raise RuntimeError('Magnetometer not enabled. Call enable_mag() first.')

        raw = self._ak8963_read_burst(self._AK8963_REG_HXL, 7)
        mx, my, mz = struct.unpack('<hhh', raw[:6])
        st2 = raw[6]

        if not (st2 & 0x08):
            pass

        if self._mag_bits == 16:
            sens = self._MAG_SENSITIVITY_16BIT
        else:
            sens = self._MAG_SENSITIVITY_14BIT

        return (mx * sens * self._mag_scale_x,
                my * sens * self._mag_scale_y,
                mz * sens * self._mag_scale_z)

    def accel_raw(self):
        """Read raw 3-axis accelerometer values.

        Returns:
            tuple: (x, y, z) raw 16-bit signed values.
        """
        raw = self._read_burst(self._REG_ACCEL_XOUT_H, 6)
        return struct.unpack('>hhh', raw)

    def gyro_raw(self):
        """Read raw 3-axis gyroscope values.

        Returns:
            tuple: (x, y, z) raw 16-bit signed values.
        """
        raw = self._read_burst(self._REG_GYRO_XOUT_H, 6)
        return struct.unpack('>hhh', raw)

    def mag_raw(self):
        """Read raw 3-axis magnetometer values.

        Returns:
            tuple: (x, y, z) raw 16-bit signed values.

        Raises:
            RuntimeError: If magnetometer has not been enabled via enable_mag().
        """
        if not self._mag_enabled:
            raise RuntimeError('Magnetometer not enabled. Call enable_mag() first.')

        raw = self._ak8963_read_burst(self._AK8963_REG_HXL, 7)
        mx, my, mz = struct.unpack('<hhh', raw[:6])
        return (mx, my, mz)

    def data_ready(self):
        """Check if new sensor data is available.

        Returns:
            bool: True when RAW_DATA_RDY_INT is set in INT_STATUS.
        """
        return bool(self._read_reg(self._REG_INT_STATUS) & 0x01)

    def set_sleep(self, sleep=True):
        """Set or clear the SLEEP bit in PWR_MGMT_1.

        Args:
            sleep: True to enter sleep mode, False to wake.
        """
        val = self._read_reg(self._REG_PWR_MGMT_1)
        if sleep:
            val |= 0x40
        else:
            val &= ~0x40
        self._write_reg(self._REG_PWR_MGMT_1, val)

    def fifo_count(self):
        """Read the number of bytes in the FIFO buffer.

        Returns:
            int: FIFO byte count (0–512).
        """
        raw = self._read_burst(self._REG_FIFO_COUNTH, 2)
        return ((raw[0] & 0x1F) << 8) | raw[1]

    def read_fifo(self):
        """Read all available data from the FIFO buffer.

        Returns:
            bytes: FIFO data (length determined by fifo_count).
        """
        count = self.fifo_count()
        if count == 0:
            return b''
        return self._read_burst(self._REG_FIFO_R_W, count)

    def enable_fifo(self, gyro=True, accel=True, temp=False):
        """Configure and enable FIFO sources.

        Args:
            gyro: Enable gyroscope data in FIFO.
            accel: Enable accelerometer data in FIFO.
            temp: Enable temperature data in FIFO.
        """
        fifo_en = ((accel & 1) << 3) | ((temp & 1) << 2) | ((gyro & 1) << 4)
        self._write_reg(self._REG_FIFO_EN, fifo_en)
        user_ctrl = self._read_reg(self._REG_USER_CTRL)
        self._write_reg(self._REG_USER_CTRL, user_ctrl | 0x40)

    def reset_fifo(self):
        """Reset the FIFO buffer by setting FIFO_RST in USER_CTRL."""
        user_ctrl = self._read_reg(self._REG_USER_CTRL)
        self._write_reg(self._REG_USER_CTRL, user_ctrl | 0x04)